Support torchbind op dispatch in python

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124239
* #123370
* __->__ #123367

We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

ydwu4 committed:
Support torchbind op dispatch in python

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"




[ghstack-poisoned]

ydwu4 commented on torch/_ops.py:
```diff
@@ -770,11 +770,10 @@ def handler(*args, **kwargs):
-                # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
-                return self.python_key_mode_table[curr_mode](*args, **kwargs)
 
-            self._dispatch_cache[key] = handler
```
When inputs are normal script object, it goes through C++ dispatcher then go back to python in `_get_dispatch`.  Previously, this logic of handling mode is never trigged because python_key_mode_table is empty ([L755](https://github.com/pytorch/pytorch/pull/123367/files#diff-9453b6b52a54783beec3dd1c60248620f61c3a524d404a188af17bbdf6be3d9eL755)) if the op is not higher order ops (higher order op dispatch completely in python so they won't go through `_get_dispatch`). 

Even though there is not implementation for these keys for custom op, it seems that there're backend fallback implementations for these keys so they can be traced fine previously. 

For TorchBindOpOverload, we'll by default implement these dispatch mode keys so the python_key_mode_table is not empty any more. when inputs are not fakified, functionalizing/tracing through them can trigger the `_get_dispatch` logic here. We therefore need to add proper fake mode support in this case.

zou3519 replied:
I'm confused, are you saying we need to invalidate the cache?

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table compared. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through higher order op's mechanism and only implements dispatch mode keys that're useful for tracing.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -971,6 +1003,19 @@ def __call__(self_, *args, **kwargs):  # noqa: B902
+        # Directly calling OverloadPacket goes into C++, which will check
+        # the schema and cause an error for torchbind op when inputs FakeScriptObject so we
+        # intercept it here and call TorchBindOpverload instead.
+        if self_._has_torchbind_op_overload:
+            # Try the overloads one by one to match an op.
+            for overload_name in self_._schemas.keys():
+                try:
+                    ret = getattr(self_, overload_name)(*args, **kwargs)
+                except RuntimeError as e:
+                    continue
+                return ret
```
Try to reuse the existing C++ logic as much as possible

ydwu4 replied:
Just checked that the overload_names are sorted already. _jit_get_operation returns a cpp func and all its overload_names in sorted order see [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/python/init.cpp#L1697-L1714). So we're OK to try in this order?

zou3519 commented on torch/_ops.py:
```diff
@@ -856,11 +855,41 @@ def tags(self):
+
+    # use `self_` to avoid naming collide with arguments that
+    # are named "self". This way, they can be called by kwargs.
+    def __call__(self_, *args, **kwargs):  # noqa: B902
+        # The path when any inputs are FakeScriptObject, we need to
+        # skip c++ dispatcher and dispatch in python like a higher order op.
+        if pytree.tree_any(
+            lambda obj: isinstance(
+                obj, torch._library.fake_class_registry.FakeScriptObject
+            ),
+            (args, kwargs),
+        ):
+            dispatch_key_set = _compute_keyset(args, kwargs, self_.non_fallthrough_keys)
+            return dispatch_in_python(
+                dispatch_key_set.highestPriorityTypeId(), self_, *args, **kwargs
+            )
```
```py
def __call__(self_, *args, **kwargs):
    if pytree.tree_any(
            lambda obj: isinstance(
                obj, torch._library.fake_class_registry.FakeScriptObject
            ),
            (args, kwargs),
    ):
        keyset = _compute_keyset(args, kwargs)
        key = highestDispatchKey(keyset)
        impl = self_._get_dispatch(key)
        if "impl_from_cpp":
            raise RuntimeError("Can't handle, needs python impl")
        return impl(*args, **kwargs)
    return self_._op(*args, **kwargs)
```

zou3519 replied:
Open question on what you need to add default fallthroughs for.

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table compared. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through higher order op's mechanism and only implements dispatch mode keys that're useful for tracing.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table compared. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through higher order op's mechanism and only implements dispatch mode keys that're useful for tracing.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table compared. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through higher order op's mechanism and only implements dispatch mode keys that're useful for tracing.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/tensor_version_op.py:
```diff
@@ -13,13 +13,13 @@
 
 @_tensor_version.py_impl(FakeTensorMode)
-def _tensor_version_fake(self):
+def _tensor_version_fake(fake_mode, self_tensor):
```
This op goes through `_get_dispatch` . We unify the handler signature of dispatch modes to handler(mode, *args, **kwargs) to be consistent with  higher order op handler and torch bind handler.

zou3519 replied:
sgtm

ydwu4 commented on torch/testing/_internal/torchbind_impls.py:
```diff
@@ -28,5 +28,15 @@ def fake_queue_push(tq, x):
+    def fake_queue_size(tq):
         return tq.size()
+
+    torch.ops._TorchScriptTesting.queue_pop.default.fallthrough(
```
We explicitly fallthrough AutogradCPU key.

zou3519 replied:
We should put this into the fallthroughs that we make in TorchBindOp's ctor

zou3519 commented on torch/_ops.py:
```diff
@@ -788,11 +786,77 @@ def tags(self):
+
+        self.non_fallthrough_keys = torch._C._dispatch_keyset_full()
+
+        _TORCH_BIND_OP_DEFAULT_FALLTHROUGH_DISPATCH_KEYS = [
```
- If the implementation is a fallthrough (as is the case for ADInplaceOrView or BackendSelect), then we shouldn't need to do anything -- I would expect `_get_dispatch` to just handle the case?
- If the implementation is a fallback for PythonDispatcher, PythonTLSSnapshot, or Autograd, then we should use a fallthrough for those. This is not completely correct; ideally we would be able to have python versions of the fallbacks that we can call.

zou3519 replied:
Sorry, let me think about this some more.

zou3519 commented on torch/_ops.py:
```diff
@@ -788,11 +786,77 @@ def tags(self):
+        return self_._op(*args, **kwargs)
+
+    def dispatch_in_python(self, args, kwargs):
+        dispatch_key_set = _compute_keyset(args, kwargs, self.non_fallthrough_keys)
```
Is this faithful to the logic at https://github.com/pytorch/pytorch/blob/main/torch/csrc/PyInterpreter.cpp#L380 ? That is, does the keyset already exclude the fallthrough keys? Or is `_get_dispatch` responsible for handling fallthroughs?

zou3519 replied:
Answer: yes it is

zou3519 commented on torch/_ops.py:
```diff
@@ -788,11 +786,77 @@ def tags(self):
+        return self_._op(*args, **kwargs)
+
+    def dispatch_in_python(self, args, kwargs):
+        dispatch_key_set = _compute_keyset(args, kwargs, self.non_fallthrough_keys)
```
This is a little tricky because there's already a source of truth for non_fallthrough_keys in C++. Compare this with the HOP case, where the entire source of truth is in Python because HOPs don't go through C++.

self.non_fallthrough_keys should come from:
- nonFallthroughKeys from C++ (the OperatorEntry has its own nonFallthroughKeys)
- This key set should then be updated for any .py_impl(non_fallthrough), .py_impl(fallthrough) that we add for the operator.

I wouldn't worry about caching this for now (also because open registration means that if we were to cache we'd need to think about cache invalidation)

zou3519 commented on torch/_ops.py:
```diff
@@ -903,13 +970,45 @@ def __call__(self_, *args, **kwargs):  # noqa: B902
+    if torch_function_called:
+        return ret
+
+    # In cpp, we do a schema matching for the arguments, and call ToIValue to
+    # to check whether the arguments are valid. But need to do similar things here
+    # and skip the ToIValue check because FakeScriptObject won't convert to IValue.
+    exceptions = {}
+    for overload_name in op.overloads():
+        try:
+            return getattr(op, overload_name)(*args, **kwargs)
+        except RuntimeError as e:
+            exceptions[overload_name] = e
+            continue
+    raise RuntimeError(
+        f"Fail to call all TorchBindOverloads of {op} with exceptions {exceptions}"
```
I read through the C++ overloading code. This doesn't feel faithful enough:
- If the operator throws a RuntimeError, then we report it differently
- what happens if we have `foo(Tensor, TensorQueue)` and `foo(Tensor, AtomicCounter)`? Do ScriptObjects have overload resolution done on them?

It sounds like:
- we need some sort of `match_schema_ignoring_fake_object -> bool` C++ function. Or maybe we just need to construct a new schema, ignoring the script object, and then use that for schema matching.
- IF ScriptObjects have overload resolution done on them, then we need to model that

zou3519 reviewed (commented):
We should try to take things more from their sources of truth:
- non_fallthrough_keys should be a combination of the C++ nonFallthroughKeys for the operator and any new py_impls (that are fallthroughs or non-fallthroughs)
- To bypass specific dispatch keys (Autograd, PythonDispatcher, BackendSelect, etc), we should check that the C++ operator didn't override them and then register a fallthrough for them. In the long term we're going to want to register a Python fallback for them.
- Left some questions on the operator overloading -- I'm not sure it's faithful enough

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented on torch/_ops.py:
```diff
@@ -782,6 +775,85 @@ def tags(self):
+        if _must_dispatch_in_python(args, kwargs):
+            # When any inputs are FakeScriptObject, we need to
+            # skip c++ dispatcher and dispatch in python through _get_dispatch of python_dispatcher.
+            return self_._dispatch_in_python(args, kwargs, self_._fallthrough_keys())
```
Compute _fallthrough_keys on the fly makes sure we don't add kernels to the dispatching logic, so that when tracing finishes, the dispatch key implementation of ops remain the same.

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -403,6 +403,10 @@ class TORCH_API OperatorHandle {
     return operatorDef_->op.hasKernelForDispatchKey(k);
   }
 
+  bool isKernelFallBackKernel(DispatchKey k) const {
```
nit: fallback is one word, so it should be "isKernelFallbackKernel"

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +775,85 @@ def tags(self):
+            DispatchKey.ADInplaceOrView,
+            DispatchKey.PythonTLSSnapshot,
+        ]
+        return [
+            key
+            for key in _DEFAULT_FALLTHROUGH_KEYS
+            if key not in self.py_kernels
+            or self.py_kernels[key] is torch.library.fallthrough_kernel
+        ]
```
This logic should be:
- if the op doesn't have a py_kernel or a C++ implementation for one of these keys, then treat it as a fallthrough.

zou3519 replied:
Also, name this to something like "mayUseFallthroughInsteadOfFallback" and add a big TODO. The problem is that we should be calling the fallback for these, but a fallthrough is almost close enough to the fallthrough in most cases that we care about that we can use a fallthrough

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +775,85 @@ def tags(self):
+            else self._dispatch_cache[dispatch_key]
+        )
+
+        if isinstance(handler, DispatchKey):
```
When is `handler` a DispatchKey?

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +775,85 @@ def tags(self):
+        if isinstance(handler, DispatchKey):
+            # fallthrough keys can be registered at runtime via torch.library.impl
+            # so need to add it to fallthrough_keys and re-dispatch.
+            if torch._C._dispatch_kernel_for_dispatch_key_is_fallthrough(  # type: ignore[attr-defined]
```
Add the type hint to `torch/_C/__init__.pyi.in` instead of the type: ignore

zou3519 commented on torch/_ops.py:
```diff
@@ -793,6 +865,14 @@ def __init__(self, qualified_op_name, op_name, op, overload_names):
         self._op = op
         self._overload_names = overload_names
         self._dir = []
+        self._schemas = {
+            overload_name: torch._C._get_schema(qualified_op_name, overload_name)
+            for overload_name in self._overload_names
+        }
```
Do you need to cache this? If not, please make this an @property

zou3519 commented on torch/_ops.py:
```diff
@@ -867,13 +957,58 @@ def __call__(self_, *args, **kwargs):  # noqa: B902
+# _jit_get_operations, which calls _get_operation_for_overload_or_packet.
+def _call_overload_packet_from_python(op: OpOverloadPacket, args, kwargs):
+    # Re-use the torch function handling logic in cpp
+    torch_function_called, ret = torch._C._maybe_call_torch_function_for_op_packet(  # type: ignore[attr-defined]
```
add to `__init__.pyi.in`

zou3519 commented on torch/_ops.py:
```diff
@@ -867,13 +957,58 @@ def __call__(self_, *args, **kwargs):  # noqa: B902
+        return found_op(*args, **kwargs)
+
+    err_msg = (
+        f"Fail to match any TorchBindOverloads of {op} with following exceptions:\n"
```
TorchBindOpOverload ?

zou3519 commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -757,6 +757,24 @@ std::pair<std::shared_ptr<Operator>, Stack> getOpWithStack(
+    const py::kwargs& kwargs) {
+  Stack stack;
+  try {
+    stack = createStackForSchema(
+        schema, std::move(args), kwargs, c10::nullopt, true);
```
er, what happens if a FakeScript object gets here?

ydwu4 replied:
We'll check it with fakeScriptObjectToIValue, which will check whether the current FakeScriptObject is an instance of the fake class corresponding to the torch class specified in the schema.

zou3519 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -873,11 +873,37 @@ struct VISIBILITY_HIDDEN tuple_slice {
   int64_t e;
 };
 
+inline IValue fakeScriptObjectToIValue(
+    const c10::FunctionSchema& schema,
+    size_t argumentPosition,
+    py::handle object) {
+  auto argument = schema.arguments().at(argumentPosition);
+  auto class_type = argument.real_type()->expect<c10::ClassType>();
+  auto fake_class_registry =
+      py::module::import("torch._library.fake_class_registry");
+  auto fake_class = fake_class_registry.attr("find_fake_class")(
+      class_type->name().value().qualifiedName());
+  if (!py::isinstance(object.attr("wrapped_obj"), fake_class)) {
+    throw schema_match_error(c10::str(
+        schema.formatTypeMismatchMsg(
+            argument,
+            friendlyTypeName(object),
+            argumentPosition,
+            py::repr(object.attr("wrapped_obj"))),
+        "\nCast error details: ",
+        argument.name(),
+        " is expected to be a FakeScriptObject of ",
+        class_type->name().value().qualifiedName()));
+  }
+  return toIValue(object, PyObjectType::get());
+}
```
I don't think we should be implementing createStackForSchema like this -- someone is going to set the allow_fake_script_boject = True and not realize that it should only be used for checking.

Instead we want a separate `bool matchesSchema(schema, args, kwargs)` function that can handle FakeScriptObject

zou3519 reviewed (commented):
Two main comments:
- the default fallthrough logic should be more faithful. What we're doing is turning a C++ fallback into a Python fallthrough. If the user has overwritten these keys, then we should respect that.
- the schema matching is suspicious. In particular, we should never be turning a FakeScriptObject into an IValue, instead what we want is some sort of new `matchesSchema -> bool` function.

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -873,6 +873,116 @@ struct VISIBILITY_HIDDEN tuple_slice {
+    const auto& argument = schema.arguments().at(arg_idx);
+    if (argument.real_type()->kind() == TypeKind::ClassType &&
+        py::isinstance(arg, fake_class_registry.attr("FakeScriptObject"))) {
+      fakeScriptObjectToIValue(schema, arg_idx, arg);
```
why are we converting the FakeScriptObject to Ivalue?

ydwu4 replied:
Yeah, changed the name and the return value to a boolean.

zou3519 commented on torch/_ops.py:
```diff
@@ -782,6 +775,101 @@ def tags(self):
+            DispatchKey.PythonTLSSnapshot,
+        ]
+
+        def _may_use_fallthrough_instead_of_fallback(key: DispatchKey):
+            _fallthrough_in_cpp_dispatcher = (
+                torch._C._dispatch_has_kernel_for_dispatch_key(self.name(), key)
+                and torch._C._dispatch_kernel_for_dispatch_key_is_fallthrough(
+                    self.name(), key
+                )
+            )
+
+            _fallthrough_in_python_dispatcher = (
+                key not in self.py_kernels
+                or self.py_kernels[key] is torch.library.fallthrough_kernel
+            )
+            return _fallthrough_in_cpp_dispatcher or _fallthrough_in_python_dispatcher
```
This is not correct -- if the key is not in self.py_kernels then this returns True, even if there's a C++ kernel for the DispatchKey. We should get rid of that condition.

ydwu4 replied:
yeah, you're right. I've changed to the new logic:
1. if has impl in cpp dispathcer, just return if the kernel is fall through kernel.
2. otherwise, if key not in py_kernels, we'll fallthrough the default ones. Or if it's py_impled with fallthrough_kernel, we can also fallthrough it.

zou3519 reviewed (approved):
I don't think `_may_use_fallthrough_instead_of_fallback` is correct -- let's try to add some more tests for that? Other than that this LGTM

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -757,6 +757,23 @@ std::pair<std::shared_ptr<Operator>, Stack> getOpWithStack(
 
+// This function is used to check if the schema is valid for the given args and
+// kwargs It skips checking script object to allow FakeScriptObject to be
+// matched. It returns the stack if the arugments matches schema and false
```
fix comment.

ydwu4 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -873,6 +873,116 @@ struct VISIBILITY_HIDDEN tuple_slice {
+      if (arg.real_type()->kind() == TypeKind::ClassType &&
+          py::isinstance(
+              cur_kwarg, fake_class_registry.attr("FakeScriptObject"))) {
+        checkFakeScriptObjectSchema(schema, i, cur_kwarg);
```
change name to `checkFakeScriptObjectMatchSchema`

ydwu4 commented on torch/_ops.py:
```diff
@@ -867,13 +966,58 @@ def __call__(self_, *args, **kwargs):  # noqa: B902
+    # The following mirrors getOpWithStack.
+    # In cpp, we do a schema matching for the arguments, and call ToIValue to
+    # to check whether the arguments are valid. But need to do similar things here
+    # and allow toIValue FakeScriptObject.
```
fix comment

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Support torchbind op dispatch in python"


We override the `__call__` method and register fake, functional, proxy default dispatch mode implementation in its python_key_mode_table. 

The idea is:
1. when inputs contains FakeScriptObject,  we dispatch it through _get_dispatch mechanism. We implement dispatch mode keys automatically in the operator's constructor.
2. when inputs are not fakified, we dispatch through the original c++ dispatcher. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
